### PR TITLE
Allow EventLoops to rethrow ThreadDeath errors

### PR DESCRIPTION
**Pull request by greenjustin**
Allows EventLoops to rethrow ThreadDeath errors.

Motivation:

Thread.stop() works by producing a ThreadDeath error in the target thread. EventLoops swallow all Throwables, which makes them effectively unkillable. This is effectively a memory leak, for our application.

Modification:

Edit the EventLoops that swallow Throwables to instead rethrow ThreadDeath errors.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@greenjustin can you please sign our ICLA as well and let me know once done ?

https://netty.io/s/icla

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java [hunk 1]
```diff
@@ -391,7 +391,11 @@ protected void run() {
                     events.increase();
                 }
             } catch (Throwable t) {
-                handleLoopException(t);
+                if (t instanceof ThreadDeath) {
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:394]: Please refactor to:

```java
try {
    ...
} catch (ThreadDeath td) {
    throw (ThreadDeath) td;
} catch (Throwable t) {
    handleLoopException(t);
}
```

Some for all other places.

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java (hunk 1 above)
Comment (greenjustin) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java]: Sure, refactored.

**Comment by greenjustin** [comment]
> @greenjustin can you please sign our ICLA as well and let me know once done ?

Yup, just got permission from legal, the ICLA is signed.

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java [hunk 2]
```diff
@@ -390,6 +390,8 @@ protected void run() {
                     //increase the size of the array as we needed the whole space for the events
                     events.increase();
                 }
+            } catch (ThreadDeath td) {
+                throw (ThreadDeath) td;
```
Comment (chrisvest) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:394]: Right below these catch clauses it says "Always handle shutdown even if the loop processing threw an exception."

Should that perhaps be moved into a finally clause? It's the same deal in all of the event loops modified in this PR.

**Review by chrisvest — COMMENTED** [review]
Had a question.

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java (hunk 2 above)
Comment (normanmaurer) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:394]: good idea @chrisvest ... @greenjustin can you please adjust ?

Comment (greenjustin) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:394]: Sure, how does this look?

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java [hunk 3]
```diff
@@ -390,19 +390,22 @@ protected void run() {
...
+                // Always handle shutdown even if the loop processing threw an exception.
+                try {
+                    if (isShuttingDown()) {
+                        closeAll();
+                        if (confirmShutdown()) {
+                            break;
+                        }
                     }
+                } catch (Throwable t) {
+                    handleLoopException(t);
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:407]: @greenjustin wouldn't you want to rethrow `ThreadDeath` here as well ? (same is true for all the other impls)

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Comment by njhill** [comment]
I wonder whether we should generalize this to `java.lang.Error`? My understanding is that these are considered fatal in general and shouldn't be handled/swallowed.

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@njhill I would be fine with that as well... @chrisvest @trustin WDYT ?

**Comment by greenjustin** [comment]
> I wonder whether we should generalize this to `java.lang.Error`? My understanding is that these are considered fatal in general and shouldn't be handled/swallowed.

I agree. Should we bother with the rethrowing code at that point? My understanding is that only Exception and Error inherit Throwable, we could just modify the "catch (Throwable t)" blocks to "catch (Exception e)".

**Comment by normanmaurer** [comment]
@greenjustin I think I would worry about this as the "user" may extend `Throwable` and so it may not be true that only `Error` and `Exception` extend it.

**Comment by greenjustin** [comment]
> @greenjustin I think I would worry about this as the "user" may extend `Throwable` and so it may not be true that only `Error` and `Exception` extend it.�

You can't extend Throwable directly, for me at least it produces a compiler error.

**Comment by greenjustin** [comment]
> > @greenjustin I think I would worry about this as the "user" may extend `Throwable` and so it may not be true that only `Error` and `Exception` extend it.�
> 
> You can't extend Throwable directly, for me at least it produces a compiler error.

Nvm, that's a weird internal-only thing. Yes I can rethrow Errors, give me a few.

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Review by njhill — APPROVED** [review]
LGTM thanks @greenjustin 

Though I'm now having second thoughts about the `Error` thing. I have seen OOMs where there's not enough direct mem for the `ByteBufAllocator` to alloc a new direct `ByteBuf` e.g. because of an unusually-sized request. Maybe it's better to handle at least OOM case so that things might otherwise continue to work/recover?

**Comment by greenjustin** [comment]
> LGTM thanks @greenjustin
> 
> Though I'm now having second thoughts about the `Error` thing. I have seen OOMs where there's not enough direct mem for the `ByteBufAllocator` to alloc a new direct `ByteBuf` e.g. because of an unusually-sized request. Maybe it's better to handle at least OOM case so that things might otherwise continue to work/recover?

Hmm...maybe, but that seems like a unique circumstance. Is it better to add Error handling logic for all OOMs just in case we might be able to handle one of them, or is it better to just crash gracefully? I'm not really sure what the consequences are for crashing in the middle of trying to handle an error.

**Comment by normanmaurer** [comment]
@njhill I think it should be just fine...

**Comment by normanmaurer** [comment]
@njhill if this happens the user should fix up their code.

**Comment by normanmaurer** [comment]
@greenjustin thanks a lot!

**Comment by greenjustin** [comment]
No worries, thanks for reviewing!

**Pull request merged**